ArcGIS feature server: implement provider test suite and fix issues

nyalldawson opened the pull request:
This PR hooks the AFS provider up to the vector provider stress/conformance test suite.

It also fixes all issues identified as a result. The most serious is that the provider was totally broken for FilterFids requests. There was also an infinite loop when iterating features through Python which was quite serious.

nyalldawson committed:
[afs] Handle layers where the id field is not named "objectid"

nyalldawson committed:
[afs] Don't list layers with no Query capability in AFS source select

Since these layers cannot be queried or added to the map and should
be added as mapserver layers instead.

nyalldawson committed:
[afs] Fix provider incorrectly overrides provider name and description
with layer name and description

nyalldawson committed:
Start implementing the provider test suite for ArcGIS Feature Server provider

nyalldawson committed:
[afs] Correctly return an invalid feature from closed iterators

nyalldawson committed:
Exempt AFS provider from certain provider tests

nyalldawson committed:
Simplify QgsVectorDataProvider::fillMinMaxCache()

nyalldawson committed:
Don't try to run provider tests which require subsetStrings on providers which don't support them

nyalldawson committed:
Avoid some detaches in AFS iterator

nyalldawson committed:
[afs] Correctly convert attribute values to field type

Avoids AFS provider returning double values for integer
and long field types

nyalldawson committed:
Better geometry memory management for afs provider

nyalldawson committed:
[afs] Remove unused code

nyalldawson committed:
[afs] Fix provider ignores FilterFids requests and returns
incorrect features

nyalldawson committed:
[afs] Fix crash when requesting invalid feature ID

nyalldawson committed:
Make feature source/provider test handle additional fields

nyalldawson committed:
[afs] Ensure that null attributes are correctly mapped for PyQGIS

nyalldawson committed:
More provider check handling of extra fields

nyalldawson committed:
Fix accidental test change

nyalldawson committed:
[afs] Don't lock shared data whilst fetching remote data

Since the fetch can be slow, there's no need to lock other
afs threads while we wait for a particular set of results to
be fetched.

nyalldawson committed:
Try to fix test which fails only on Travis

nyalldawson committed:
[afs] Fix/optimise handling of filter rect feature requests

Before a filter rect request would usually force fetching every
single feature from the server before the request could be
complete.

Instead, if a filter rect is passed we first obtain a list
from the server of matching features within this rect, and
then iterate only over those.

Fixes broken (multi-minute hang) identify tool use on AFS
layers.

nyalldawson committed:
[afs] Add unit test for non OBJECTID named id fields

nirvn commented:
I've just tested this PR, it's a huge improvement on large datasets, fantastic work.

nyalldawson committed:
[afs] Support interruption requests on feature iterator

Results in a huge increase in responsiveness when working with
large AFS layers

nyalldawson committed:
[afs] Correctly advertise SelectAtId capability

Big speedup for attribute table loading with provider

nyalldawson committed:
[afs] Don't request intersecting features from server if all
features are already locally cached

nyalldawson committed:
Avoid redundant error

nyalldawson committed:
[afs] More responsive cancelation for feature requests

nyalldawson committed:
Add a bounding box intersection test to QgsGeometry

We only previously had methods for exact intersections - this
commit adds a new QgsGeometry.boundingBoxIntersects() method
which can be used to test if just the bounding boxes of
geometries/rectangles intersect.

It's fast, and doesn't care about invalid geometries (unlike
the exact intersects checks)

nyalldawson committed:
[afs] Use faster bounding box checks where possible

nyalldawson committed:
Fix links are doubled in metadata widget

nyalldawson committed:
[afs] Auto populate layer metadata from provider metadata

nyalldawson committed:
[afs] Add unit test for metadata

nyalldawson committed:
Spelling
